sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/stream/DataStreams.java

```diff
@@ -186,14 +186,16 @@ public DataStreamDecoder(Coder<T> coder, PrefetchableIterator<ByteString> inputS
      * ByteString} in the underlying {@link ByteString} {@link Iterator iterator} and decoding
      * elements till at the next boundary.
      */
-    public List<T> decodeFromChunkBoundaryToChunkBoundary() {
+    public WeightedList<T> decodeFromChunkBoundaryToChunkBoundary() {
       inbound.currentStream = inputByteStrings.next().newInput();
       inbound.position = 0;
       try {
         InputStream previousStream = inbound.currentStream;
-        List<T> rvals = new ArrayList<>();
+        WeightedList<T> rvals = new WeightedList<>(new ArrayList<>(), 0L);
```

Comment:
You could alternatively create the list of all elements as before, then a create the weighted list with all the bytes read, rather than tracking the bytes read for each element.

Reply:
This can be easily done. Same question:

Hmm that's true, although I don't think the overhead of keeping an extra counter adds too much.

Do you think weight per element isn't useful? I can think it would come handy in a few optimizations like avoid recomputing the sizes if an element was removed, but those are further optimizations / I didn't want to change more than necessary.